Regression: dict | None or dict fallback breaks type inference for nested dicts after PR #19249

<!--
If you're not sure whether what you're experiencing is a mypy bug, please see the "Question and Help" form instead.
Please consider:

- checking our common issues page: https://mypy.readthedocs.io/en/stable/common_issues.html
- searching our issue tracker: https://github.com/python/mypy/issues to see if it's already been reported
- asking on gitter chat: https://gitter.im/python/typing
-->

**Bug Report**

<!--
If you're reporting a problem with a specific library function, the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues

If the project you encountered the issue in is open source, please provide a link to the project.
-->

After upgrading mypy past [PR #19249](https://github.com/python/mypy/pull/19249), we observed new errors related to dictionary fallback using or expressions when value types are unions of nested dicts and optional types.

This appears to be a regression in type inference, as the same code passed in previous versions of mypy.

**To Reproduce**
```python 
from typing import Union, Optional, Any

P = dict[str, Union[Optional[str], dict[str, Optional[str]]]]

def f(x: P) -> None:
    pass

def g(x: dict[str, Any] | None, s: str | None) -> None:
    f(x or {'x': s})  # ❌ Error in master branch, passes in latest mypy release (1.17.0)
```

mypy playground Gist: https://mypy-play.net/?mypy=master&python=3.12&gist=c4a5711c815d342425983850913896fe

**Actual Behavior**

<!-- What went wrong? Paste mypy's output. -->

```
main.py:9: error: Argument 1 to "f" has incompatible type "dict[str, Any] | dict[str, str | None]"; expected "dict[str, str | dict[str, str | None] | None]"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

<!-- Include as many relevant details about the environment you experienced the bug in -->

- Mypy version used: mypy 1.18.0+dev.abb1cc787455501e43f344c8af2de7d0d262e5ef
- Mypy command-line flags:
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: 3.11

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -6004,6 +6004,10 @@ class ExpressionChecker(ExpressionVisitor[Type], ExpressionCheckerSharedApi):
     def _combined_context(self, ty: Type | None) -> Type | None:
         ctx_items = []
         if ty is not None:
+            if has_any_type(ty):
+                # HACK: Any should be contagious, `dict[str, Any] or <x>` should still
+                # infer Any in x.
+                return ty
             ctx_items.append(ty)
         if self.type_context and self.type_context[-1] is not None:
             ctx_items.append(self.type_context[-1])

--- a/test-data/unit/check-inference-context.test
+++ b/test-data/unit/check-inference-context.test
@@ -1530,3 +1530,13 @@ def check3(use: bool, val: str) -> "str | Literal[False]":
 def check4(use: bool, val: str) -> "str | bool":
     return use and identity(val)
 [builtins fixtures/tuple.pyi]
+
+[case testDictAnyOrLiteralInContext]
+from typing import Union, Optional, Any
+
+def f(x: dict[str, Union[str, None, int]]) -> None:
+    pass
+
+def g(x: Optional[dict[str, Any]], s: Optional[str]) -> None:
+    f(x or {'x': s})
+[builtins fixtures/dict.pyi]

